File: go/116_populating_next_right_pointers_in_each_node/main.js
// code
/**
 * // Definition for a Node.
 * function Node(val,left,right,next) {
 *    this.val = val;
 *    this.left = left;
 *    this.right = right;
 *    this.next = next;
 * };
 */
/**
 * @param {Node} root
 * @return {Node}
 */
let connect = function(root) {
  if (null == root) return root;

  let pre = root;
  let curr;
  while (pre.left) {
    curr = pre;
    while (curr) {
      curr.left.next = curr.right;
      if (curr.next) curr.right.next = curr.next.left;
      curr = curr.next;
    }
    pre = pre.left;
  }
  
  return root;
};

File: go/116_populating_next_right_pointers_in_each_node/main.go
package main

type Node struct {
	Val   int
	Left  *Node
	Right *Node
	Next  *Node
}

// code
/**
 * Definition for a Node.
 * type Node struct {
 *     Val int
 *     Left *Node
 *     Right *Node
 *     Next *Node
 * }
 */

func connect(root *Node) *Node {
	node := root
	for node != nil {
		iter := node

		node = node.Left
		if node == nil {
			break
		}

		for iter != nil {
			iter.Left.Next = iter.Right
			if iter.Next != nil {
				iter.Right.Next = iter.Next.Left
			}
			iter = iter.Next
		}
	}

	return root
}
